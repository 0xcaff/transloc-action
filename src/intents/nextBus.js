// @flow
import type { DialogflowApp, OptionItem } from "actions-on-google";
import type { Arrival, Route, Stop, RouteStops } from "transloc-api";
import { getArrivals, getRoutes, getStops } from "transloc-api";

import type { Coords } from "../utils";
import {
  coordsToPosition,
  distance,
  simplifyDuration,
  lowestCost,
  makeMap,
  timeUntil,
  mustGet,
  pluralizeByCount,
  pluralizeDo,
  sortByDistance
} from "../utils";

import logger from "../logger";
import { now } from "../now";

export const NEXT_BUS_INTENT = "bus.next";
export const NEXT_BUS_LOCATION_INTENT = "bus.next.location";
export const NEXT_BUS_OPTION_INTENT = "bus.next.option";

const agencies = ["643"];

export const FROM_ARGUMENT = "from";

const getFromArg = (app: DialogflowApp): ?string =>
  app.getArgument(FROM_ARGUMENT);

export const TO_ARGUMENT = "to";

const getToArg = (app: DialogflowApp): ?string => app.getArgument(TO_ARGUMENT);

type ArrivalWithRoute = Arrival & { +route: Route };
type ArrivalWithRouteStops = Arrival & { +route: RouteWithStop };
type RouteWithStop = Route & RouteStops;

const getArrivalsWithRoute = async (
  stop: Stop
): Promise<ArrivalWithRoute[]> => {
  const arrivalsResponse = await getArrivals({ agencies, stop_id: stop.id });
  logger.info({ arrivalsResponse }, "getArrivals response");
  const { arrivals } = arrivalsResponse;

  const routesResponse = await getRoutes({ agencies });
  logger.info({ routesResponse }, "getRoutes response");
  const { routes } = routesResponse;
  const routesMap: Map<number, Route> = makeMap(routes);

  return arrivals.map(arrival => ({
    ...arrival,
    get route(): Route {
      return mustGet(routesMap, arrival.route_id);
    }
  }));
};

const stitchRouteStops = (
  arrivals: ArrivalWithRoute[],
  routeStops: RouteStops[]
): $ReadOnlyArray<ArrivalWithRouteStops> => {
  const routeStopsMap: Map<number, RouteStops> = makeMap(routeStops);

  return arrivals.map(
    (arrival: ArrivalWithRoute): ArrivalWithRouteStops =>
      (({
        ...arrival,
        get route(): RouteWithStop {
          const route = arrival.route;
          const routeStops = mustGet(routeStopsMap, arrival.route_id);

          return {
            ...route,
            ...routeStops
          };
        }
      }: any): ArrivalWithRouteStops)
  );
};

type OptionKey = {
  id: number,
  type: string
};

const handleUnknownStops = (
  potentialStops: Stop[],
  type: string,
  stopName: string,
  app: DialogflowApp
): void => {
  const failedToFindStopMsg: string = `I couldn't find a stop named "${stopName}."`;
  const sortedPotentialStops: Stop[] = sortByDistance(
    potentialStops,
    stopName,
    stop => stop.name
  );

  if (!app.hasSurfaceCapability(app.SurfaceCapabilities.SCREEN_OUTPUT)) {
    app.tell(failedToFindStopMsg);
    return;
  }

  // Display List of Stops
  const list = app.buildList("Stops").addItems(
    sortedPotentialStops.map((stop: Stop): OptionItem => {
      const option = app.buildOptionItem(
        JSON.stringify(({ id: stop.id, type }: OptionKey))
      );

      option.setTitle(stop.name);
      option.setDescription(stop.description);

      return option;
    })
  );

  app.askWithList(failedToFindStopMsg, list);
};

export const nextBus = async (app: DialogflowApp): Promise<void> => {
  logger.info("handling next bus intent");

  const to = getToArg(app);
  const from = getFromArg(app);
  logger.info({ to, from }, "arguments");

  const { stops, routes } = await getStops({ agencies, include_routes: !!to });
  logger.info({ stops, routes }, "getStops response");

  const fromStop: ?Stop = resolveFromStop(app, from, stops);
  if (!fromStop) {
    return;
  }
  logger.info({ fromStop }, "resolved from stop");

  const arrivals = await getArrivalsWithRoute(fromStop);
  if (!arrivals.length) {
    app.tell(`There are no buses arriving at ${fromStop.name}.`);
    return;
  }

  let filteredArrivals: $ReadOnlyArray<ArrivalWithRoute> = arrivals;
  const optionStop: ?Stop = getStopFromOption(stops, TO_ARGUMENT, app);
  const matchingStop: ?Stop = findMatchingStop(to, stops);
  let resolvedToStop: ?Stop = optionStop || matchingStop;

  if (to && !matchingStop) {
    handleUnknownStops(stops, TO_ARGUMENT, to, app);
    return;
  }

  if (resolvedToStop) {
    logger.info({ resolvedToStop }, "found matching to stop");
    if (!routes) {
      logger.error({ routes }, `The API response didn't include routes`);
      app.tell("Something went wrong.");

      return;
    }

    const stitchedRoutes = stitchRouteStops(arrivals, routes);
    filteredArrivals = stitchedRoutes.filter(arrival =>
      arrival.route.stops.find(id => id === resolvedToStop.id)
    );
  }

  logger.info({ filteredArrivals }, "filter arrivals");

  createResponse(app, fromStop, resolvedToStop, filteredArrivals);
};

const createResponse = (
  app: DialogflowApp,
  from: Stop,
  to: ?Stop,
  inputArrivals: $ReadOnlyArray<ArrivalWithRoute>
): void => {
  const currentTime = now();
  const arrivals = inputArrivals
    .slice()
    .filter(arrival => currentTime < arrival.timestamp);

  // Sort arrivals by timestamp in ascending order (smallest first).
  arrivals.sort((a, b) => a.timestamp - b.timestamp);

  const topArrivals = arrivals.slice(0, 5);

  const arrivalsInfo = topArrivals.map(({ route, timestamp }) => {
    const duration = simplifyDuration(timeUntil(timestamp));

    return { duration, routeName: route.long_name };
  });

  const textArrivals = arrivalsInfo
    .map(
      ({ duration, routeName }) =>
        `${routeName} in ${duration.count} ${pluralizeByCount(
          duration.unit,
          duration.count
        )}`
    )
    .join("; ");

  let predicate: string = "";
  if (!to) {
    predicate = `arriving at ${from.name}`;
  } else {
    predicate = `traveling from ${from.name} to ${to.name}`;
  }

  if (arrivals.length) {
    app.tell(
      `The following ${pluralizeByCount("bus", arrivals.length)} ${pluralizeDo(
        arrivals.length
      )} ${predicate}. ${textArrivals}.`
    );
  } else {
    app.tell(`There are no buses ${predicate}.`);
  }
};

const findNearestStop = (to: Coords, stops: Stop[]): ?Stop =>
  lowestCost(stops, stop => {
    const devicePosition = coordsToPosition(to);
    const distanceToStop = distance(devicePosition, (stop.position: any));

    return distanceToStop;
  });

const findMatchingStop = (query: ?string, stops: Stop[]): ?Stop => {
  if (!query) {
    return;
  }

  const normalizedFrom = query.toLowerCase().trim();
  const stop = stops.find(
    element => element.name.toLowerCase().trim() === normalizedFrom
  );

  return stop;
};

// Gets the nearest stop to the current location, requesting permission if
// needed.
const getStopByLocation = (stops: Stop[], app: DialogflowApp): ?Stop => {
  if (!app.isPermissionGranted()) {
    logger.info("requesting location permission");

    const fromLocation = app.askForPermission(
      "To find the nearest stop",
      app.SupportedPermissions.DEVICE_PRECISE_LOCATION
    );

    if (fromLocation === null) {
      throw new TypeError(`Failed to ask for location permission.`);
    }

    // After this, the location is collected and the intent with the event
    // actions_intent_PERMISSION is triggered.
    return;
  }

  const location = app.getDeviceLocation();
  logger.info({ location }, "location permission granted");
  if (!location) {
    app.tell(`I couldn't get your location.`);
    logger.info("failed to get location");
    return;
  }

  const { coordinates: deviceCoordinates } = location;
  const nearestStop = findNearestStop(deviceCoordinates, stops);
  logger.info({ nearestStop }, "resolved nearest stop");

  if (nearestStop === null) {
    app.tell("There aren't any stops. I'm not sure what to do.");
    return;
  }

  return nearestStop;
};

const getStopFromOption = (
  stops: Stop[],
  argumentType: string,
  app: DialogflowApp
): ?Stop => {
  logger.info({ argumentType }, "getting from option");
  const selectedOption = app.getSelectedOption();
  logger.info({ selectedOption }, "selected option");
  if (!selectedOption) {
    return;
  }

  // Try to get from argument.
  const { id, type } = JSON.parse(selectedOption);
  if (type !== argumentType) {
    return;
  }

  const stop = stops.find((stop: Stop) => stop.id === id);
  if (!stop) {
    return;
  }

  logger.info({ argumentType, stop }, "got stop from option");
  return stop;
};

// Tries to find the nearest stop to `from`. If not provided, uses the
// current location.
const resolveFromStop = (
  app: DialogflowApp,
  from: ?string,
  stops: Stop[]
): ?Stop => {
  const optionStop = getStopFromOption(stops, FROM_ARGUMENT, app);
  if (optionStop) {
    return optionStop;
  }

  if (!from) {
    // From not provided, try to use device location.
    return getStopByLocation(stops, app);
  }

  const stop = findMatchingStop(from, stops);
  logger.info({ stop }, "found matching stop");
  if (!stop) {
    handleUnknownStops(stops, FROM_ARGUMENT, from, app);
    return;
  }

  return stop;
};
