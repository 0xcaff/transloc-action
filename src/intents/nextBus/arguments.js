// @flow
import type { DialogflowApp } from "actions-on-google";

export const FROM_ARGUMENT = "from";

export const getFromArg = (app: DialogflowApp): ?string =>
  app.getArgument(FROM_ARGUMENT);

export const TO_ARGUMENT = "to";

export const getToArg = (app: DialogflowApp): ?string =>
  app.getArgument(TO_ARGUMENT);
