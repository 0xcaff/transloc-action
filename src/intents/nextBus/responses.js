// @flow
import type { Stop, RouteStops } from "transloc-api";
import type { DialogflowApp, List, OptionItem } from "actions-on-google";
import type { ArrivalWithRoute } from "../../data";
import {
  pluralizeByCount,
  pluralizeDo,
  simplifyDuration,
  sortByDistance,
  stringifyDuration,
  timeUntil
} from "../../utils";
import { now } from "../../now";
import type { OptionKey, OptionType } from "./option";
import logger from "../../logger";
import { getArrivalsWithRoute, stitchRouteStops } from "../../data";

export const findAndShowArrivals = async (
  app: DialogflowApp,
  from: Stop,
  maybeTo: ?Stop,
  routes: ?(RouteStops[])
): Promise<void> => {
  const arrivals = await findArrivals(app, from, maybeTo, routes);
  if (!arrivals) {
    return;
  }

  sendResponse(app, from, maybeTo, arrivals);
};

// Find arrivals traveling from a stop to another stop. Returns null if
// an invalid API response was received.
const findArrivals = async (
  app: DialogflowApp,
  from: Stop,
  maybeTo: ?Stop,
  routes: ?(RouteStops[])
): Promise<?$ReadOnlyArray<ArrivalWithRoute>> => {
  const arrivals = await getArrivalsWithRoute(from);

  if (maybeTo) {
    const to: Stop = maybeTo;

    if (!routes) {
      logger.error({ routes }, `The API response didn't include routes`);
      app.tell("Something went wrong.");
      return null;
    }

    const stitchedRoutes = stitchRouteStops(arrivals, routes);
    return stitchedRoutes.filter(arrival =>
      arrival.route.stops.find(id => id === to.id)
    );
  }

  return arrivals;
};

// Arranges and sorts information to respond to the next bus query.
const sendResponse = (
  app: DialogflowApp,
  from: Stop,
  to: ?Stop,
  inputArrivals: $ReadOnlyArray<ArrivalWithRoute>
): void => {
  const currentTime = now();

  // Ignore Arrivals In Past
  const arrivals = inputArrivals
    .slice()
    .filter(arrival => currentTime < arrival.timestamp);

  // Sort arrivals by timestamp in ascending order (smallest first).
  arrivals.sort((a, b) => a.timestamp - b.timestamp);

  const topArrivals = arrivals.slice(0, 5);

  const arrivalsInfo = topArrivals.map(({ route, timestamp }) => {
    const duration = simplifyDuration(timeUntil(timestamp));

    return { duration, routeName: route.long_name };
  });

  const textArrivals = arrivalsInfo
    .map(
      ({ duration, routeName }) =>
        `${routeName} in ${stringifyDuration(duration)}`
    )
    .join("; ");

  let predicate: string = "";
  if (!to) {
    predicate = `arriving at ${from.name}`;
  } else {
    predicate = `traveling from ${from.name} to ${to.name}`;
  }

  if (arrivals.length) {
    app.tell(
      `The following ${pluralizeByCount("bus", arrivals.length)} ${pluralizeDo(
        arrivals.length
      )} ${predicate}. ${textArrivals}.`
    );
  } else {
    app.tell(`There are no buses ${predicate}.`);
  }
};

export const handleUnknownStops = (
  potentialStops: Stop[],
  type: OptionType,
  stopName: string,
  app: DialogflowApp
): void => {
  const failedToFindStopMsg: string = `I couldn't find a stop named "${stopName}."`;
  const sortedPotentialStops: Stop[] = sortByDistance(
    potentialStops,
    stopName,
    stop => stop.name
  );

  if (!app.hasSurfaceCapability(app.SurfaceCapabilities.SCREEN_OUTPUT)) {
    app.tell(failedToFindStopMsg);
    return;
  }

  const list = displayStopsList(app, type, sortedPotentialStops);

  app.askWithList(failedToFindStopMsg, list);
};

// Creates a list of formatted stops.
export const displayStopsList = (
  app: DialogflowApp,
  type: OptionType,
  potentialStops: Stop[]
): List =>
  app.buildList("Stops").addItems(
    potentialStops.map((stop: Stop): OptionItem => {
      const option = app.buildOptionItem(
        JSON.stringify(({ id: stop.id, type }: OptionKey))
      );

      option.setTitle(stop.name);
      option.setDescription(stop.description);

      return option;
    })
  );
