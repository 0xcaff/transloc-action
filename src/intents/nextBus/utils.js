// @flow
import type { Stop } from "transloc-api";

import type { Coordinates } from "actions-on-google";
import { coordsToPosition, distance, lowestCost } from "../../utils";

// Find the nearest stop the the specified co-ordinates.
export const findNearestStop = (to: Coordinates, stops: Stop[]): ?Stop =>
  lowestCost(stops, stop => {
    const devicePosition = coordsToPosition(to);
    const distanceToStop = distance(devicePosition, (stop.position: any));

    return distanceToStop;
  });

// Finds the stop which most closely matches the query. If there isn't a
// close match, returns null.
export const findMatchingStop = (query: ?string, stops: Stop[]): ?Stop => {
  if (!query) {
    return;
  }

  const normalizedFrom = query.toLowerCase().trim();
  const stop = stops.find(
    element => element.name.toLowerCase().trim() === normalizedFrom
  );

  return stop;
};
