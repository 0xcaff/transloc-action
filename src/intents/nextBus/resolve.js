// @flow
import type { Stop } from "transloc-api/lib/endpoints/stops";
import { DialogflowApp } from "actions-on-google";
import type { OptionKey, OptionType } from "./option";
import { FROM_OPTION_TYPE, TO_OPTION_TYPE } from "./option";
import { findMatchingStop, findNearestStop } from "./utils";
import logger from "../../logger";
import { sortByDistance } from "../../utils";
import { handleUnknownStops } from "./responses";
import { getFromArg, getToArg } from "./arguments";
import type { Result } from "../../result";

// Find a stop matching the specified query. If none can be found, a list of
// stops is shown to the user.
const findOrRequestMatchingStop = (
  app: DialogflowApp,
  query: string,
  stops: Stop[],
  type: OptionType
): Result<Stop> => {
  // From was provided, try finding a matching stop.
  const stop = findMatchingStop(query, stops);
  logger.info({ stop }, `found matching "${(type: any)}" stop`);
  if (!stop) {
    // No stop matches, return suggestions.
    askWithStopList(app, type, query, stops);
    return { type: "DELEGATING" };
  }

  return { type: "SUCCESS", value: stop };
};

// Displays a list of stops, sorted by closest match to furthest match.
const askWithStopList = (
  app: DialogflowApp,
  type: OptionType,
  stopName: string,
  potentialStops: Stop[]
): void => {
  const sortedPotentialStops: Stop[] = sortByDistance(
    potentialStops,
    stopName,
    stop => stop.name
  );

  handleUnknownStops(sortedPotentialStops, type, stopName, app);
};

// Tries to find the stop most closely named to `to`.
export const resolveToStop = (
  app: DialogflowApp,
  stops: Stop[]
): Result<Stop> => {
  const to: ?string = getToArg(app);
  if (!to) {
    return { type: "EMPTY" };
  }

  return findOrRequestMatchingStop(app, to, stops, TO_OPTION_TYPE);
};

// Tries to find the stop named most closely to `from`. If not provided,
// uses the current location. If not provided, requests the current location.
export const resolveFromStop = (
  app: DialogflowApp,
  stops: Stop[]
): Result<Stop> => {
  const from: ?string = getFromArg(app);

  if (!from) {
    // From not provided, try to use device location.
    return getStopByLocation(app, stops);
  }

  // From was provided, try finding a matching stop.
  return findOrRequestMatchingStop(app, from, stops, FROM_OPTION_TYPE);
};

// Get's the stop nearest to the current location. If the location isn't
// available, requests it.
const getStopByLocation = (app: DialogflowApp, stops: Stop[]): Result<Stop> => {
  const location = app.getDeviceLocation();
  if (location && location.coordinates) {
    // The location was provided previously, let's use it.

    const { coordinates: deviceCoordinates } = location;
    const nearestStop = findNearestStop(deviceCoordinates, stops);

    if (nearestStop) {
      return { type: "SUCCESS", value: nearestStop };
    }
  }

  // There was no suitable location, let's ask for permission.
  logger.info("requesting location permission");

  const fromLocation = app.askForPermission(
    "To find the nearest stop",
    app.SupportedPermissions.DEVICE_PRECISE_LOCATION
  );

  if (fromLocation === null) {
    throw new TypeError(`Failed to ask for location permission.`);
  }

  // After this, the location is collected and the intent with the event
  // actions_intent_PERMISSION is triggered.

  return { type: "DELEGATING" };
};

export const getStopFromOption = (
  app: DialogflowApp,
  option: OptionKey,
  type: OptionType,
  stops: Stop[]
): Result<Stop> => {
  if (option.type !== type) {
    return { type: "EMPTY" };
  }

  const stop = getStopById(option.id, stops);
  if (!stop) {
    logger.warn({ option, stops }, "missing selected stop");
    app.tell(`The selected stop couldn't be found.`);
    return { type: "DELEGATING" };
  }

  return { type: "SUCCESS", value: stop };
};

export const getStopById = (stopId: number, stops: Stop[]): ?Stop =>
  stops.find(stop => stop.id === stopId);
