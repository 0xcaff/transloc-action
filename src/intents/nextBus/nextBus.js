// @flow
import type { DialogflowApp } from "actions-on-google";
import type { Stop } from "transloc-api";

import { getStops } from "../../data";

import logger from "../../logger";
import { agencies } from "../../data/agencies";
import { findAndShowArrivals } from "./responses";
import type { Result } from "./resolve";
import { convertResult, resolveFromStop, resolveToStop } from "./resolve";
import { FROM_STOP_KEY, storeLocationContext } from "./context";

// The intent handler for the next bus intent.
export const nextBus = async (app: DialogflowApp): Promise<void> => {
  logger.info("handling next bus intent");

  const { stops, routes } = await getStops({
    agencies,
    include_routes: true
  });

  const maybeFromStop: Result<Stop> = resolveFromStop(app, stops);
  if (maybeFromStop.type === "DELEGATING") {
    logger.info("delegating from stop resolution");
    // Either a request for clarification or a location request was performed.
    return;
  }

  if (maybeFromStop.type === "EMPTY") {
    throw new TypeError(`Didn't expect the from stop to be empty.`);
  }

  const fromStop: Stop = maybeFromStop.value;
  logger.info({ fromStop }, "resolved from stop");

  // Set "from" Context
  storeLocationContext(app, FROM_STOP_KEY, fromStop);

  const maybeToStop: Result<Stop> = resolveToStop(app, stops);
  if (maybeToStop.type === "DELEGATING") {
    logger.info("delegating to stop resolution");
    return;
  }

  const convertedMaybeToStop: ?Stop = convertResult(maybeToStop);

  return findAndShowArrivals(app, fromStop, convertedMaybeToStop, routes);
};
