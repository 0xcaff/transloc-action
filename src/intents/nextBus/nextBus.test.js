// @flow
jest.mock("../../now");

import { FROM_ARGUMENT, TO_ARGUMENT } from "./arguments";
import type { DeviceLocation } from "actions-on-google";
import { DialogflowApp } from "actions-on-google";
import { nextBus } from "./nextBus";
import { MockDialogflowApp } from "../../mockDialogflowApp";

const location: DeviceLocation = {
  coordinates: { latitude: 43.082978, longitude: -77.677036 },
  address: "105 Lomb Memorial Dr",
  zipCode: "14623",
  city: "Rochester, NY"
};

it(`requests the location when "from" is not provided`, async () => {
  const app = new MockDialogflowApp();
  app.permissionGranted = false;

  await nextBus((app: any));

  expect(app.response).toMatchSnapshot();
  expect(app.contextOut).toMatchSnapshot();
});

it(`uses a previously authorized location if provided`, async () => {
  const app = new MockDialogflowApp();
  app.deviceLocation = location;
  app.permissionGranted = false;

  await nextBus((app: any));

  expect(app.response).toMatchSnapshot();
  expect(app.contextOut).toMatchSnapshot();
});

it("works with an explicit source location", async () => {
  const app = new MockDialogflowApp(new Map([["from", "Gleason Circle"]]));
  app.permissionGranted = false;

  await nextBus((app: any));

  expect(app.response).toMatchSnapshot();
  expect(app.contextOut).toMatchSnapshot();
});

it("works with an explicit source and destination", async () => {
  const app = new MockDialogflowApp(
    new Map([
      [FROM_ARGUMENT, "Gleason Circle"],
      [TO_ARGUMENT, "Park Point South"]
    ])
  );
  app.permissionGranted = false;

  await nextBus((app: any));

  expect(app.response).toMatchSnapshot();
  expect(app.contextOut).toMatchSnapshot();
});

it("works with a location based source and explicit destination", async () => {
  const app = new MockDialogflowApp(
    new Map([[TO_ARGUMENT, "Park Point South"]])
  );
  app.permissionGranted = true;
  app.deviceLocation = location;

  await nextBus((app: any));

  expect(app.response).toMatchSnapshot();
  expect(app.contextOut).toMatchSnapshot();
});

it("works on a route without any buses", async () => {
  const app = new MockDialogflowApp(
    new Map([[FROM_ARGUMENT, "Gleason Circle"], [TO_ARGUMENT, "Target"]])
  );

  await nextBus((app: any));

  expect(app.response).toMatchSnapshot();
  expect(app.contextOut).toMatchSnapshot();
});

it("should not return a list on incompatible surfaces", async () => {
  const app = new MockDialogflowApp(
    new Map([[FROM_ARGUMENT, "Gleason Circle"], [TO_ARGUMENT, "Unknown"]])
  );
  const realApp: DialogflowApp = (app: any);
  app.surfaceCapabilities.add(realApp.SurfaceCapabilities.AUDIO_OUTPUT);

  await nextBus(realApp);

  expect(app.response).toMatchSnapshot();
  expect(app.contextOut).toMatchSnapshot();
});

it("should return a list on compatible surfaces", async () => {
  const app = new MockDialogflowApp(
    new Map([
      [FROM_ARGUMENT, "Gleason Circle"],
      [TO_ARGUMENT, "Barnes and Noble"]
    ])
  );
  const realApp: DialogflowApp = (app: any);
  app.surfaceCapabilities.add(realApp.SurfaceCapabilities.AUDIO_OUTPUT);
  app.surfaceCapabilities.add(realApp.SurfaceCapabilities.SCREEN_OUTPUT);

  await nextBus(realApp);

  expect(app.response).toMatchSnapshot();
  expect(app.contextOut).toMatchSnapshot();
});

// TODO: should resolve a query with a selected to and a location based from
// TODO: should resolve a query with a given from without any busses
// TODO: Should resolve a query which returns a single bus
// TODO: Test Failure Case With Specified From and To
