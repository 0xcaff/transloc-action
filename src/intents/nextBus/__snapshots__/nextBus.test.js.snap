// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`requests the location when "from" is not provided 1`] = `
Object {
  "permission": Object {
    "context": "To find the nearest stop",
    "dialogState": undefined,
    "permission": "DEVICE_PRECISE_LOCATION",
  },
  "type": "ASK_FOR_PERMISSION",
}
`;

exports[`requests the location when "from" is not provided 2`] = `Map {}`;

exports[`should not return a list on incompatible surfaces 1`] = `
Object {
  "response": "I couldn't find a stop named \\"Unknown.\\"",
  "type": "TELL",
}
`;

exports[`should not return a list on incompatible surfaces 2`] = `
Map {
  "from" => Object {
    "lifespan": 0,
    "name": "from",
    "parameters": Object {
      "stopId": 4197446,
    },
  },
}
`;

exports[`should return a list on compatible surfaces 1`] = `
Object {
  "list": List {
    "items": Array [
      OptionItem {
        "description": undefined,
        "image": undefined,
        "optionInfo": Object {
          "key": "{\\"id\\":4197454,\\"type\\":\\"to\\"}",
          "synonyms": Array [],
        },
        "title": "Barnes & Noble",
      },
      OptionItem {
        "description": undefined,
        "image": undefined,
        "optionInfo": Object {
          "key": "{\\"id\\":4209568,\\"type\\":\\"to\\"}",
          "synonyms": Array [],
        },
        "title": "Park Point North",
      },
      OptionItem {
        "description": undefined,
        "image": undefined,
        "optionInfo": Object {
          "key": "{\\"id\\":4212884,\\"type\\":\\"to\\"}",
          "synonyms": Array [],
        },
        "title": "Corner Lot B",
      },
      OptionItem {
        "description": undefined,
        "image": undefined,
        "optionInfo": Object {
          "key": "{\\"id\\":4177296,\\"type\\":\\"to\\"}",
          "synonyms": Array [],
        },
        "title": "Marketplace Mall",
      },
      OptionItem {
        "description": undefined,
        "image": undefined,
        "optionInfo": Object {
          "key": "{\\"id\\":4209570,\\"type\\":\\"to\\"}",
          "synonyms": Array [],
        },
        "title": "Park Point South",
      },
      OptionItem {
        "description": undefined,
        "image": undefined,
        "optionInfo": Object {
          "key": "{\\"id\\":4197446,\\"type\\":\\"to\\"}",
          "synonyms": Array [],
        },
        "title": "Gleason Circle",
      },
      OptionItem {
        "description": undefined,
        "image": undefined,
        "optionInfo": Object {
          "key": "{\\"id\\":4209574,\\"type\\":\\"to\\"}",
          "synonyms": Array [],
        },
        "title": "Target",
      },
      OptionItem {
        "description": undefined,
        "image": undefined,
        "optionInfo": Object {
          "key": "{\\"id\\":4177292,\\"type\\":\\"to\\"}",
          "synonyms": Array [],
        },
        "title": "Colony Manor",
      },
      OptionItem {
        "description": undefined,
        "image": undefined,
        "optionInfo": Object {
          "key": "{\\"id\\":4197450,\\"type\\":\\"to\\"}",
          "synonyms": Array [],
        },
        "title": "Perkins Green",
      },
      OptionItem {
        "description": undefined,
        "image": undefined,
        "optionInfo": Object {
          "key": "{\\"id\\":4177288,\\"type\\":\\"to\\"}",
          "synonyms": Array [],
        },
        "title": "The Province",
      },
      OptionItem {
        "description": undefined,
        "image": undefined,
        "optionInfo": Object {
          "key": "{\\"id\\":4177298,\\"type\\":\\"to\\"}",
          "synonyms": Array [],
        },
        "title": "Wegmans",
      },
      OptionItem {
        "description": undefined,
        "image": undefined,
        "optionInfo": Object {
          "key": "{\\"id\\":4197480,\\"type\\":\\"to\\"}",
          "synonyms": Array [],
        },
        "title": "Racquet Club",
      },
      OptionItem {
        "description": undefined,
        "image": undefined,
        "optionInfo": Object {
          "key": "{\\"id\\":4177290,\\"type\\":\\"to\\"}",
          "synonyms": Array [],
        },
        "title": "Residence Halls",
      },
      OptionItem {
        "description": undefined,
        "image": undefined,
        "optionInfo": Object {
          "key": "{\\"id\\":4209580,\\"type\\":\\"to\\"}",
          "synonyms": Array [],
        },
        "title": "Jefferson Plaza",
      },
      OptionItem {
        "description": undefined,
        "image": undefined,
        "optionInfo": Object {
          "key": "{\\"id\\":4197466,\\"type\\":\\"to\\"}",
          "synonyms": Array [],
        },
        "title": "Wal-Mart",
      },
      OptionItem {
        "description": undefined,
        "image": undefined,
        "optionInfo": Object {
          "key": "{\\"id\\":4177316,\\"type\\":\\"to\\"}",
          "synonyms": Array [],
        },
        "title": "Riverknoll South",
      },
      OptionItem {
        "description": undefined,
        "image": undefined,
        "optionInfo": Object {
          "key": "{\\"id\\":4177314,\\"type\\":\\"to\\"}",
          "synonyms": Array [],
        },
        "title": "UC West",
      },
      OptionItem {
        "description": undefined,
        "image": undefined,
        "optionInfo": Object {
          "key": "{\\"id\\":4209582,\\"type\\":\\"to\\"}",
          "synonyms": Array [],
        },
        "title": "RIT Inn",
      },
      OptionItem {
        "description": undefined,
        "image": undefined,
        "optionInfo": Object {
          "key": "{\\"id\\":4197484,\\"type\\":\\"to\\"}",
          "synonyms": Array [],
        },
        "title": "Global Village",
      },
      OptionItem {
        "description": undefined,
        "image": undefined,
        "optionInfo": Object {
          "key": "{\\"id\\":4209572,\\"type\\":\\"to\\"}",
          "synonyms": Array [],
        },
        "title": "South Town Plaza",
      },
      OptionItem {
        "description": undefined,
        "image": undefined,
        "optionInfo": Object {
          "key": "{\\"id\\":4209576,\\"type\\":\\"to\\"}",
          "synonyms": Array [],
        },
        "title": "Regal Cinemas",
      },
      OptionItem {
        "description": undefined,
        "image": undefined,
        "optionInfo": Object {
          "key": "{\\"id\\":4209578,\\"type\\":\\"to\\"}",
          "synonyms": Array [],
        },
        "title": "Jefferson & Scutti",
      },
      OptionItem {
        "description": undefined,
        "image": undefined,
        "optionInfo": Object {
          "key": "{\\"id\\":4212882,\\"type\\":\\"to\\"}",
          "synonyms": Array [],
        },
        "title": "Lot B",
      },
      OptionItem {
        "description": undefined,
        "image": undefined,
        "optionInfo": Object {
          "key": "{\\"id\\":4209584,\\"type\\":\\"to\\"}",
          "synonyms": Array [],
        },
        "title": "Tech Park / John St.",
      },
    ],
    "title": "Stops",
  },
  "type": "ASK_WITH_LIST",
}
`;

exports[`should return a list on compatible surfaces 2`] = `
Map {
  "from" => Object {
    "lifespan": 0,
    "name": "from",
    "parameters": Object {
      "stopId": 4197446,
    },
  },
}
`;

exports[`uses a previously authorized location if provided 1`] = `
Object {
  "response": "The following buses are arriving at Gleason Circle. Perkins - Park Point in 11 minutes; Colony Manor / The Province in 16 minutes; RIT Inn in 16 minutes; Racquet Club in 27 minutes; Perkins - Park Point in 32 minutes.",
  "type": "TELL",
}
`;

exports[`uses a previously authorized location if provided 2`] = `
Map {
  "from" => Object {
    "lifespan": 0,
    "name": "from",
    "parameters": Object {
      "stopId": 4197446,
    },
  },
}
`;

exports[`works on a route without any buses 1`] = `
Object {
  "response": "There are no buses traveling from Gleason Circle to Target.",
  "type": "TELL",
}
`;

exports[`works on a route without any buses 2`] = `
Map {
  "from" => Object {
    "lifespan": 0,
    "name": "from",
    "parameters": Object {
      "stopId": 4197446,
    },
  },
}
`;

exports[`works with a location based source and explicit destination 1`] = `
Object {
  "response": "The following buses are traveling from Gleason Circle to Park Point South. Perkins - Park Point in 11 minutes; Perkins - Park Point in 32 minutes.",
  "type": "TELL",
}
`;

exports[`works with a location based source and explicit destination 2`] = `
Map {
  "from" => Object {
    "lifespan": 0,
    "name": "from",
    "parameters": Object {
      "stopId": 4197446,
    },
  },
}
`;

exports[`works with an explicit source and destination 1`] = `
Object {
  "response": "The following buses are traveling from Gleason Circle to Park Point South. Perkins - Park Point in 11 minutes; Perkins - Park Point in 32 minutes.",
  "type": "TELL",
}
`;

exports[`works with an explicit source and destination 2`] = `
Map {
  "from" => Object {
    "lifespan": 0,
    "name": "from",
    "parameters": Object {
      "stopId": 4197446,
    },
  },
}
`;

exports[`works with an explicit source location 1`] = `
Object {
  "response": "The following buses are arriving at Gleason Circle. Perkins - Park Point in 11 minutes; Colony Manor / The Province in 16 minutes; RIT Inn in 16 minutes; Racquet Club in 27 minutes; Perkins - Park Point in 32 minutes.",
  "type": "TELL",
}
`;

exports[`works with an explicit source location 2`] = `
Map {
  "from" => Object {
    "lifespan": 0,
    "name": "from",
    "parameters": Object {
      "stopId": 4197446,
    },
  },
}
`;
